rust/opendp/src/interactive.rs

```diff
@@ -0,0 +1,274 @@
+            return fallible!(DomainMismatch, "wrong query input domain")
+        } else if measurement.output_domain.as_ref() != &self.output_domain {
+            return fallible!(DomainMismatch, "wrong query output domain")
+        } else if measurement.input_metric.as_ref() != &self.input_metric {
+            return fallible!(DomainMismatch, "wrong query input metric")
+        } else if measurement.output_measure.as_ref() != &self.output_measure {
+            return fallible!(DomainMismatch, "wrong query output metric")
+        }
+        Ok(())
+    }
+
+    /// Checks that there is adequate budget in this Queryable state.
+    fn check_budget(&self, privacy_relation: &PrivacyRelation<MI, MO>, d_out_query: &MO::Distance) -> Fallible<()> {
+        privacy_relation.eval(&self.d_in_budget, d_out_query)?;
+        if d_out_query > &self.d_out_budget {
+            return fallible!(FailedRelation, "not enough budget")
+        }
+        Ok(())
+    }
+
+    /// Updates this Queryable state by consuming the given amount of budget.
+    fn update(self, d_out_query: &MO::Distance) -> Fallible<Self> {
+        Ok(Self { d_out_budget: self.d_out_budget.sub(d_out_query)?, ..self })
+    }
+
+    /// Processes a query, generating a new Queryable state.
+    fn transition(self, (measurement, d_out_query): &AcQuery<DI, DO, MI, MO>) -> Fallible<(Self, DO::Carrier)>
+        where MO::Distance: Clone + MeasureDistance {
+        self.check_types(measurement)?;
+        self.check_budget(&measurement.privacy_relation, d_out_query)?;
+        let res = measurement.function.eval(&self.data)?;
+        let new = self.update(d_out_query)?;
+        Ok((new, res))
+    }
+}
+type AcQuery<DI, DO, MI, MO> = (Measurement<DI, DO, MI, MO>, <MO as Measure>::Distance);
+type AcQueryable<DI, DO, MI, MO> = Queryable<AcState<DI, DO, MI, MO>, AcQuery<DI, DO, MI, MO>, <DO as Domain>::Carrier>;
+type AcMeasurement<DI, DO, MI, MO> = InteractiveMeasurement<DI, DO, MI, MO, AcState<DI, DO, MI, MO>, AcQuery<DI, DO, MI, MO>>;
+
+pub fn make_adaptive_composition<DI, DO, MI, MO>(
+    input_domain: DI,
+    output_domain: DO,
+    input_metric: MI,
+    output_measure: MO,
+    d_in_budget: MI::Distance,
+    d_out_budget: MO::Distance,
+) -> AcMeasurement<DI, DO, MI, MO>
+    where DI: 'static + Domain,
+          DI::Carrier: Clone,
+          DO: 'static + Domain,
+          MI: 'static + Metric,
+          MI::Distance: 'static + MetricDistance + Clone,
+          MO: 'static + Measure,
+          MO::Distance: 'static + MeasureDistance + Clone {
+    AcMeasurement::new(
+        input_domain.clone(),
+        AllDomain::new(),
+        Function::new(enclose!((input_domain, input_metric, output_measure, d_in_budget, d_out_budget), move |arg: &DI::Carrier| -> AcQueryable<DI, DO, MI, MO> {
+            AcQueryable::new(
+                // TODO: Remove these clones and have the Queryable use refs? (Also remove Clone trait bounds.)
```

Comment:
You can drop input_domain from enclose!, because you only use input_domain in the closure. Since you only use input_domain in the closure, it is fine for the closure to consume it.

Reply:
To drop these clones I feel like you'd want a lifetime bound on AcState based on the lifetime of AcMeasurement. To do that I think you'd have to somehow pass Function a reference to the measurement it is a part of, to get a reference to AcMeasurement's input_domain (line 140), output_domain (141), input metric (148) and output measure (149). Not going to happen.

Reply:
Put a different way, the desired behavior requires a struct that contains a reference to itself.

Reply:
Thanks, I knew there were issues here, this was the cleanest I could work out. I think this is a good section for us to revisit in a pair programming session.